flink-runtime/src/test/java/org/apache/flink/runtime/concurrent/FutureUtilsTest.java

```diff
@@ -824,6 +825,63 @@ private void testFutureContinuation(
         Assert.assertTrue(continuationFuture.isDone());
     }
 
+    @Test
+    public void testHandleExceptionWithCompletedFuture() {
+        final CompletableFuture<String> future = CompletableFuture.completedFuture("foobar");
+        final CompletableFuture<String> handled =
+                FutureUtils.handleException(future, Exception.class, exception -> "handled");
+        assertEquals("foobar", handled.join());
+    }
+
+    @Test
+    public void testHandleExceptionWithNormalCompletion() {
+        final CompletableFuture<String> future = new CompletableFuture<>();
+        final CompletableFuture<String> handled =
+                FutureUtils.handleException(future, Exception.class, exception -> "handled");
+        future.complete("foobar");
+        assertEquals("foobar", handled.join());
+    }
+
+    @Test
+    public void testHandleExceptionWithMatchingExceptionallyCompletedFuture() {
+        final CompletableFuture<String> future = new CompletableFuture<>();
+        final CompletableFuture<String> handled =
+                FutureUtils.handleException(
+                        future, UnsupportedOperationException.class, exception -> "handled");
+        future.completeExceptionally(new UnsupportedOperationException("foobar"));
+        assertEquals("handled", handled.join());
+    }
+
+    @Test
+    public void testHandleExceptionWithNotMatchingExceptionallyCompletedFuture() {
+        final CompletableFuture<String> future = new CompletableFuture<>();
+        final CompletableFuture<String> handled =
+                FutureUtils.handleException(
+                        future, UnsupportedOperationException.class, exception -> "handled");
+        future.completeExceptionally(new IllegalArgumentException("foobar"));
+        final CompletionException completionException =
+                assertThrows(CompletionException.class, handled::join);
+        assertTrue(completionException.getCause() instanceof IllegalArgumentException);
+        assertEquals("foobar", completionException.getCause().getMessage());
```

Comment:
Might be simplified to `assertEquals(completionException.getCause(), failureCause)`.

Reply:
Alternatively, we could think about adding a `Matcher` to `FlinkMatchers` that checks the type and message of an exception.